client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -1155,23 +1013,199 @@ private int getCloseConnectionTimeout() {
     }
 
     /**
-     * Wrapped connection handler used in the {@link #connect()} method. Allows adding an additional handler.
+     * Encapsulates a connection attempt (with possible retries).
      */
-    private static class ConnectMethodConnectionHandler implements Handler<AsyncResult<HonoConnection>> {
+    private class ConnectionAttempt {
 
+        private final ProtonClientOptions clientOptions;
         private final Handler<AsyncResult<HonoConnection>> connectionHandler;
-        private final Handler<AsyncResult<HonoConnection>> additionalHandler;
+        private final AtomicInteger connectAttempts = new AtomicInteger(0);
+        private final AtomicBoolean cancelled = new AtomicBoolean();
+
+        private Long reconnectTimerId;
+
+        ConnectionAttempt(final ProtonClientOptions clientOptions, final Handler<AsyncResult<HonoConnection>> connectionHandler) {
+            this.clientOptions = clientOptions;
+
+            this.connectionHandler = ar -> {
+                if (ar.failed()) {
+                    clearState();
+                }
+                currentConnectionAttempt.compareAndSet(this, null);
+                connectionHandler.handle(ar);
+                deferredConnectionCheckHandler.setConnectionAttemptFinished(ar);
+            };
+        }
+
+        /**
+         * Starts the connection attempt.
+         * Must be called on the vert.x context created for the connection attempt.
+         *
+         * @param isReconnect if {@code true}, the connection attempt is started after the configured minimum
+         *            reconnect delay.
+         * @return {@code true} if the attempt was started, {@code false} if there already is an ongoing attempt.
+         */
+        public boolean start(final boolean isReconnect) {
+            if (!currentConnectionAttempt.compareAndSet(null, this)) {
+                // there already is another ongoing attempt
+                return false;
+            }
```

Comment:
shouldn't this better be done in HonoConnectionImpl.connect ? There seems to be no need to first instantiate a ConnectionAttempt and then figuring out that it should be discarded, is there?

Reply:
> shouldn't this better be done in HonoConnectionImpl.connect ?

I prefer to make both kind of updates to the `currentConnectionAttempt` reference close to one another in the `ConnectionAttempt` inner class.

> There seems to be no need to first instantiate a ConnectionAttempt

I want to use `currentConnectionAttempt.compareAndSet` here, so an instance of `ConnectionAttempt` is indeed needed.
The (usually exceptional) case that it is created unnecessarily should be no big deal here.

Reply:
I see